File: 820-find-eventual-safe-states/find-eventual-safe-states.py
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        n=len(graph)
        unsafe=[0]*n
        vis=[0]*n
        pathvis=[0]*n
        def dfs(vertex):
            vis[vertex]=1
            pathvis[vertex]=1
            for adj in graph[vertex]:
                if pathvis[adj]:
                    unsafe[vertex]=1
                    pathvis[vertex]=0
                    return True
                if not vis[adj]:
                    if dfs(adj)==True:
                        unsafe[vertex]=1
                        pathvis[vertex]=0
                        return True
                elif unsafe[adj]:
                        unsafe[vertex]=1
                        pathvis[vertex]=0
                        return True
            pathvis[vertex]=0
            return False
        for i in range(n):
            if not vis[i]:
                dfs(i)
        ans=[]
        for i in range(n):
            if unsafe[i]==0:
                ans.append(i)
        return ans
            

File: 820-find-eventual-safe-states/find-eventual-safe-states.cpp
class Solution {
public:

    bool dfs(vector<vector<int>>& graph,vector<int>& visited,vector<int> & inrecursion,vector<int>& safe, int current_node){
        if (safe[current_node]!=-1){
            return safe[current_node];
        }
        vector<int>& neighbours=graph[current_node];
        visited[current_node]=1;
        if (neighbours.size()==0){
            safe[current_node]=1;
            return true;
        }
        inrecursion[current_node]=1;
        for (int neighbour : neighbours){
            if (!visited[neighbour]){
                if (dfs(graph,visited,inrecursion,safe,neighbour)==false){
                    safe[current_node]=0;
                    return false;
                }
            }
            else if (inrecursion[neighbour]){
                safe[current_node]=0;
                return false;
            }
        }
        inrecursion[current_node]=0;
        safe[current_node]=1;
        return true;
    }

    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int n=graph.size();
        vector<int> safe(n,-1);
        vector<int> visited(n,0);
        vector<int> inrecursion(n,0);
        for (int i=0;i<n;i++){
            if (visited[i]==0){
                dfs(graph,visited,inrecursion,safe,i);
            }
        }
        vector<int> result;
        for (int i=0;i<n;i++){
            if (safe[i]==1){
                result.push_back(i);
            }
        }
        return result;

    }
};